Document/0x05e-Testing-Cryptography.md

```diff
@@ -1,16 +1,70 @@
 
+KeyStore and KeyChain provides APIs for storing and using keys (behind the scene, KeyChain API uses KeyStore system). These systems allow to administer the full lifecycle of the cryptographic keys. Requirements and guidance for implementation of cryptographic key management can be found in [Key Management Cheat Sheet](https://owasp.org/www-project-cheat-sheets/cheatsheets/Key_Management_Cheat_Sheet.html "Key Management Cheat Sheet"). We can identify following phases:
+
+- generating of a key
+- using of a key
+- storing of a key
+- archiving of a key
+- deleting of a key
+
+> Please note that storing of a key is analyzed in the chapter "[Testing Data Storage](0x05d-Testing-Data-Storage.md)".
+
+These phases are managed by Keystore/KeyChain system. However how the system works depends on how application developer used it. Then, you should focus on functions which are used by application developer. You should identify and verify following functions:
+
+- [Key generation](0x05e-Testing-Cryptography.md#key-generation)
+- [Random number generation](0x05e-Testing-Cryptography.md#random-number-generation)
+
+You can go further and verify how good is a key by performing [key attestation](0x05e-Testing-Cryptography.md#key-attestation).
+
+Apps that target modern API levels, went through the following changes:
+
+- For Android 7.0 (API level 24) and above [the Android Developer blog shows that](https://android-developers.googleblog.com/2016/06/security-crypto-provider-deprecated-in.html "Security provider Crypto deprecated in Andorid N"):
+  - It is recommended to stop specifying a security provider. Instead, always use a patched security provider.
+  - The support for the `Crypto` provider has dropped and the provider is deprecated. The same applies to its `SHA1PRNG` for secure random.
+- For Android 8.1 (API level 27) and above the [Developer Documentation](https://developer.android.com/about/versions/oreo/android-8.1 "Cryptography updates") shows that:
+  - Conscrypt, known as `AndroidOpenSSL`, is preferred above using Bouncy Castle and it has new implementations: `AlgorithmParameters:GCM` , `KeyGenerator:AES`, `KeyGenerator:DESEDE`, `KeyGenerator:HMACMD5`, `KeyGenerator:HMACSHA1`, `KeyGenerator:HMACSHA224`, `KeyGenerator:HMACSHA256`, `KeyGenerator:HMACSHA384`, `KeyGenerator:HMACSHA512`, `SecretKeyFactory:DESEDE`, and `Signature:NONEWITHECDSA`.
+  - You should not use the `IvParameterSpec.class` anymore for GCM, but use the `GCMParameterSpec.class` instead.
```

Comment:
IT would be good to detail that a bit, as there are a few problems with GCM (maybe we have it elsewhere with the guide):
the Nonce should NOT be repeated for the same key (as it is no IV), so there is a statemachine involved and/lor you need to change the key every time as well (e.g. use envelope-encryption). and there are limits to ohw many times you can use it.

Reply:
Added to #1689